timemachine/fe/interaction_group_traj.py

```diff
@@ -0,0 +1,113 @@
+from typing import Callable
+
+import numpy as np
+from jax import jit
+from jax import numpy as jnp
+from jax import vmap
+from numpy.typing import NDArray as Array
+
+from timemachine.potentials import nonbonded
+from timemachine.potentials.jax_utils import distance2
+
+Position = Array
+Param = Array
+Box = Array
+Energy = float
+PairFxn = Callable[[Position, Position, Param, Param, Box], Energy]
+
+
+# example pair function
+def nb_pair_fxn(x_a, x_b, param_a, param_b, box):
+    beta = 2.0
+    cutoff = 1.2
+
+    # alchemical distance
+    r2 = distance2(x_a, x_b, box)
+    w_offset = param_b[3] - param_a[3]
+    r = jnp.sqrt(r2 + w_offset**2)
+
+    # TM reaction field
+    q_prod = param_a[0] * param_b[0]
+    e_q = nonbonded.switched_direct_space_pme(r, q_prod, beta, cutoff)
+
+    # Lennard-Jones
+    sig = nonbonded.combining_rule_sigma(param_a[1], param_b[1])
+    eps = nonbonded.combining_rule_epsilon(param_a[2], param_b[2])
+    e_lj = nonbonded.lennard_jones(r, sig, eps)
+
+    return jnp.where(r < cutoff, e_q + e_lj, 0.0)
+
+
+# brute-force pre-computed neighborlist
+@jit
+def env_mask_within_cutoff(x_env, x_lig, box, cutoff):
+    """result[i] = True if any distance(x_env[i], y) < cutoff for y in x_lig"""
+
+    def d2_others(x_i, x_others):
+        d2_ij = vmap(distance2, (None, 0, None))(x_i, x_others, box)
+        return jnp.where(d2_ij <= cutoff**2, d2_ij, jnp.inf)
+
+    def within_cutoff(point):
+        return jnp.any(d2_others(point, x_lig) < cutoff**2)
+
+    return vmap(within_cutoff)(x_env)
+
+
+class InteractionGroupTraj:
+    def __init__(self, pair_fxn: PairFxn, xs: Array, box_diags: Array, ligand_idxs: Array, env_idxs: Array, cutoff=1.2):
+        r"""support [U_ig(x; params) for x in traj]
+
+        where U_ig = \sum_i \sum_j pair_fxn(||x_j - x_j||; params_i, params_j)
+
+        (with i summing over ligand_idxs, j summing over env_idxs, and pair_fxn(r) == 0 when r >= cutoff)
+        """
+        self.n_frames = len(xs)
+        self.pair_fxn = pair_fxn
+        self.ligand_idxs = ligand_idxs
+        self.all_env_idxs = env_idxs
+        num_lig, num_env = len(ligand_idxs), len(env_idxs)
+
+        # apply pair_fxn(x_i, x_j, param_i, param_j, box)
+        axes_a = (0, None, 0, None, None)
+        axes_b = (None, 0, None, 0, None)
+        self.all_pairs_fxn = vmap(vmap(pair_fxn, axes_a), axes_b)
```

Comment:
Can you modify compute_Us to take in the pair_fxn and set up all_pairs_fxn there? This would allow for more flexibility (i.e. if we want to compute q or LJ on the same traj).

Reply:
Good idea : and it also simplifies serialization (since now only arrays and floats (not jax functions) appear as instance attributes). 

Modified from `_ = traj.compute_Us(params)` to `compute_Us = traj.make_U_fxn(pair_fxn); _ = compute_Us(params)` in https://github.com/proteneer/timemachine/pull/1383/commits/8f5c064b93b7100dc0a82b071cdcf53a884de70a